File: python/2015/03/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def update(a, p):
    if a == ">":
        return [p[0]+1, p[1]]
    elif a == "<":
        return [p[0]-1, p[1]]
    elif a == "^":
        return [p[0], p[1]+1]
    elif a == "v":
        return [p[0], p[1]-1]
    else:
        pass

def read_input():
    return u.read_lines(file)

def part1():
    A = read_input()
    k = 0
    p = [0,0]
    L = []
    L.append(p)
    for a in A[0]:
        p = update(a, p)
        if p not in L: 
            L.append(p)
    print("Part 1:",len(L))

def part2():
    A = read_input()
    k = 0
    p = [0,0]
    pp = [0,0]
    L = []
    L.append(p)
    for i,a in enumerate(A[0]):
        if i % 2 == 0:
            p = update(a, p)
            if p not in L: 
                L.append(p)
        else:
            pp = update(a, pp)
            if pp not in L: 
                L.append(pp)
    print("Part 2:",len(L))
    

part1()
part2()

File: c/2015/03/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

void update(char c, int32_t* x, int32_t* y) {
    if (c == '>') {
        *x += 1;
    } else if (c == '<') {
        *x -= 1;
    } else if (c == '^') {
        *y += 1;
    } else if (c == 'v') {
        *y -= 1;
    }
}

bool inStack(POINT* pts, int16_t size, POINT p) {
    for (int16_t i = 0; i < size; i++) {
        POINT q = pts[i];
        if (comparePoints(p, q)) return true;
    }
    return false;
}

void part1() {
    ARR_STRING* in = readInput(FILE_NAME);
    printf("%d\n",in->length);
    char* s = in->strings[0];
    int64_t k = 0;
    POINT* pts = (POINT*) malloc(1024 * sizeof(POINT));
    int32_t x = 0, y = 0;
    int16_t index = 0;
    int16_t size = 1024;
    for (int32_t i = 0; i < strlen(s)+1; i++) {
        POINT p = {.x=x, .y=y, .z=0};
        if (index + 1 >= size) {
            size += 1024;
            pts = realloc(pts, size * sizeof(POINT));
        } 
        if (!inStack(pts, index, p)) {
            pts[index++] = p;
        }
        update(s[i], &x, &y);
    }
    k = index;
    printf("Part 1: %ld\n", k);
    free(pts);
    freeArrString(in);
}

void part2() {
    ARR_STRING* in = readInput(FILE_NAME);
    char* s = in->strings[0];
    int64_t k = 0;
    POINT* pts = (POINT*) malloc(1024 * sizeof(POINT));
    int32_t x = 0, y = 0, xx = 0, yy = 0;
    int16_t index = 0;
    int16_t size = 1024;
    for (int32_t i = 0; i < strlen(s)+1; i+=2) {
        POINT p1 = {.x=x, .y=y, .z=0};
        POINT p2 = {.x=xx, .y=yy, .z=0};
        
        if (index + 2 >= size) {
            size += 1024;
            pts = realloc(pts, size * sizeof(POINT));
        } 
        if (!inStack(pts, index, p1)) {
            pts[index++] = p1;
        }
        if (!inStack(pts, index, p2)) {
            pts[index++] = p2;
        }
        update(s[i], &x, &y);
        update(s[i+1], &xx, &yy);
    }
    k = index;
    printf("Part 2: %ld\n", k);
    free(pts);
    freeArrString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}
